### ISSUE TITLE
Weird exception when mocking interface with SignalRHubAttribute applied

### ISSUE THREAD

**Issue by kimbirkelund** [devlooped/moq#1308]
Hi

First of all thanks for a nice framework 👍

I've run into a weird problem. The following tests shows the problem, with the first test exposing the problem:

```
public class Test
{
    private readonly ITestOutputHelper _outputHelper;
    public Test(ITestOutputHelper outputHelper) { _outputHelper = outputHelper; }

    [Fact]
    public void Should_throw_NullReferenceException_when_mocking_wrapping_interface()
    {
        var mock = new Mock<INest<IThrow>>();
        var setup = mock.Setup(m => m.Get());
        var exception = setup.Invoking(s => s.Returns(new Mock<IThrow>().Object))
                             .Should().Throw<NullReferenceException>()
                             .Which;

        _outputHelper.WriteLine(exception.ToString());
    }

    [Fact]
    public void Should_not_throw_NullReferenceException_when_mocking_wrapping_interface()
    {
        var mock = new Mock<INest<IDoNotThrow>>();
        var setup = mock.Setup(m => m.Get());
        setup.Returns(new Mock<IDoNotThrow>().Object); // no exception occurs
    }

    [Fact]
    public void Should_not_throw_NullReferenceException_when_mocking_wrapping_interface_using_the_other_attribute()
    {
        var mock = new Mock<INest<IDoNotThrowEither>>();
        var setup = mock.Setup(m => m.Get());
        setup.Returns(new Mock<IDoNotThrowEither>().Object); // no exception occurs
    }

    [Fact]
    public void Should_not_throw_when_mocking_directly()
    {
        var mock = new Mock<IThrow>();
        var setup = mock.Setup(m => m.Get());
        setup.Returns("value"); // no exception occurs
    }
}

[CausesNullReferenceException(null)]
public interface IThrow
{
    string Get();
}

[CausesNullReferenceException(new string[0])]
public interface IDoNotThrow
{
    string Get();
}

[DoesNotCauseNullReferenceException(null)]
public interface IDoNotThrowEither
{
    string Get();
}

public interface INest<T>
{
    T Get();
}

[AttributeUsage(AttributeTargets.Interface, Inherited = false)]
public sealed class CausesNullReferenceExceptionAttribute : Attribute
{
    public CausesNullReferenceExceptionAttribute(string[]? p) { }
}

[AttributeUsage(AttributeTargets.Interface)]
public sealed class DoesNotCauseNullReferenceExceptionAttribute : Attribute
{
    public DoesNotCauseNullReferenceExceptionAttribute(string[]? p) { }
}
```

The only case the exception occurs is if the `AttributeUsageAttribute.Inherited` property is explicitly set to `false` (even though that is the default value) and the attribute has an array parameter that is `null`.

This is the exception (using LINQPad):
<img width="933" alt="image" src="https://user-images.githubusercontent.com/1884017/206176755-277dc7ae-c682-46f4-ba8f-4e27ccbdb011.png">

I'm using Moq version 4.18.3 and .NET 7. Also in use is Xunit and FluentAssertions.

**Comment by stakx** [comment]
Thanks for the error report, this will likely boil down to a problem in the DynamicProxy library. I'll investigate shortly and, if Moq indeed isn't the culprit, forward a modified report to their repository.

**Comment by stakx** [comment]
@kimbirkelund, I cannot reproduce any test failures. I had to change the following lines of code in order to get a runnable test:

```diff
-        var exception = setup.Invoking(s => s.Returns(new Mock<IThrow>().Object))
-                             .Should().Throw<NullReferenceException>()
-                             .Which;
+        var exception = Assert.Throws<NullReferenceException>(() => s.Returns(new Mock<IThrow>().Object));
```

Could you please re-post minimally complete repro code (ideally with no package references other than to Xunit and Moq) that has at least one failing test?

**Comment by kimbirkelund** [comment]
Interesting. I've attached a project that shows the issue when running `dotnet test` on my machine. 

And just to avoid any misunderstanding: all tests **_pass_** if the issue is repro'ed. The test with the assertion demonstrates the problem by **_catching_** the exception that shouldn't be thrown. I apologize if that wasn't clear by me original post.

I've further attached the result of running `dotnet --info` on my machine.

Let me know if you need anything further.

[dotnetinfo.txt](https://github.com/moq/moq4/files/10203008/dotnetinfo.txt)
[BugRepro.zip](https://github.com/moq/moq4/files/10202999/BugRepro.zip)

**Comment by stakx** [comment]
> The test with the assertion demonstrates the problem by catching the exception that shouldn't be thrown.

Ah, I see. That's a somewhat unusual way to go about it :-) but in that case I can repro it. Will take another look!

**Comment by kimbirkelund** [comment]
Sorry that wasn't clear. I thought it was the most precise way to show where the exception was thrown and what kind it was.

**Comment by stakx** [comment]
No big deal. I'm usually expecting a test failure to signal where something goes wrong. Also, your test name `Should_throw_NullReferenceException_when_mocking_wrapping_interface` suggested that a `NullReferenceException` would be the expected outcome, so when the test passed I didn't even look more closely at it.

Never mind though, I know what I need to look at now. :-)

**Comment by kimbirkelund** [comment]
Yeah, that wasn't really thought through :-)

**Comment by stakx** [comment]
@kimbirkelund, this is indeed a bug in DynamicProxy. I've forwarded your bug report with simplified test cases, see https://github.com/castleproject/Core/issues/637.

**Comment by stakx** [comment]
@kimbirkelund, I've just released Moq version 4.18.4, which contains a bugfix that should resolve your issue.

**Comment by kimbirkelund** [comment]
Good to hear. Looking forward to get it.

Thanks for your hard work 👍
